crates/jstz_proto/src/executor/fa_withdraw.rs

```diff
@@ -0,0 +1,346 @@
+        let hash = value.hash().map_err(|_| Error::InvalidTicketType)?;
+        Ok(Self { value, hash })
+    }
+}
+
+type OutboxMessageId = String;
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct FaWithdrawalReceiptContent {
+    pub source: PublicKeyHash,
+    pub outbox_message_id: OutboxMessageId,
+}
+
+impl FaWithdrawalReceiptContent {
+    pub fn to_http_body(&self) -> HttpBody {
+        Some(String::as_bytes(&json!(&self).to_string()).to_vec())
+    }
+}
+
+#[derive(Display, Debug, Error, From)]
+pub enum FaWithdrawError {
+    InvalidTicketInfo,
+    ProxySmartFunctionCannotBeSource,
+}
+
+fn create_fa_withdrawal_message(
+    routing_info: &RoutingInfo,
+    ticket: FA2_1Ticket,
+) -> Result<OutboxMessage> {
+    let receiver_pkh = routing_info.receiver.to_base58();
+    let destination = Contract::Originated(routing_info.proxy_l1_contract.clone());
+    let message = OutboxMessage::new_withdrawal_message(
+        &Contract::try_from(receiver_pkh).unwrap(),
+        &destination,
+        ticket,
+        WITHDRAW_ENTRYPOINT,
+    )?;
+    Ok(message)
+}
+
+// Deducts `amount` from the ticket balance of `ticket_owner` for `ticket.hash`
+// and pushes a withdraw outbox message to the outbox queue, returning the outbox
+// message id.
+fn withdraw_from_ticket_owner(
+    rt: &mut impl HostRuntime,
+    tx: &mut Transaction,
+    ticket_owner: &Address,
+    routing_info: &RoutingInfo,
+    amount: Amount,
+    ticket: Ticket,
+) -> Result<OutboxMessageId> {
+    TicketTable::sub(rt, tx, ticket_owner, &ticket.hash, amount)?;
+    let message = create_fa_withdrawal_message(routing_info, ticket.value)?;
+    tx.queue_outbox_message(rt, message)?;
+    // TODO: https://linear.app/tezos/issue/JSTZ-113/implement-outbox-message-id
+    // Implement outbox message id
+    Ok("".to_string())
+}
+
+/// Execute the [FaWithdrawal] request by deducting ticket balance from `source`` and
+/// pushing a withdraw message to the outbox queue. `proxy_l1_contract` is expected to
+/// implement the %withdraw entrypoint. See /jstz/contracts/examples/fa_ticketer/fa_ticketer.mligo.
+///
+/// Fails if:
+/// * Source account has insufficient funds
+/// * Outbox queue is full
+/// * Amount is zero
+fn fa_withdraw(
+    rt: &mut impl HostRuntime,
+    tx: &mut Transaction,
+    source: &Address,
+    fa_withdrawal: FaWithdrawal,
+) -> Result<FaWithdrawalReceiptContent> {
+    if fa_withdrawal.amount == 0 {
+        Err(Error::ZeroAmountNotAllowed)?
+    }
+    let FaWithdrawal {
+        amount,
+        routing_info,
+        ticket_info,
+    } = fa_withdrawal;
+    let ticket = ticket_info.to_ticket(amount)?;
+    let outbox_message_id =
+        withdraw_from_ticket_owner(rt, tx, source, &routing_info, amount, ticket)?;
+    Ok(FaWithdrawalReceiptContent {
+        source: source.clone(),
+        outbox_message_id,
+    })
+}
+
+/// Execute the [FaWithdrawal] request atomically. See [fa_withdraw].
```

Comment:
stylistic nit: But wouldn't it make more sense to have this in an `impl` for `FaWithdraw` 
e.g. `FaWithdraw::execute(rt, tx, source, gas_limit)`